### PR TITLE
Implement CIDR blocklisting

### PR DESCRIPTION
**Pull request by gregsfortytwo**
This PR lets users specify a CIDR range when blocklisting, and propagates
that range throughout the OSDMap blocklisting functionality.

To facilitate this, it also cleans up some of the blocklist handling in CephFS, and
implements blocklisting unit tests.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by gregsfortytwo** [comment]
ceph test api

**Comment by gregsfortytwo** [comment]
@batrick @neha-ojha @vshankar Obviously later than I'd planned, but I haven't done this bit twiddling in a while and my first working solution was too ugly to share. o_0 Please take a look!

There are a *few* teuthology tests that invoke blocklisting so I'm gonna go over those and see if any are possible to extend to test the ranges, but it's gonna be rough in the lab with the limited machines. And documentation incoming shortly.

**Comment by gregsfortytwo** [comment]
https://pulpito.ceph.com/gregf-2021-12-02_21:15:39-rados-wip-cidr-blocklist-distro-basic-smithi for a few runs of existing rados blocklist tests
https://pulpito.ceph.com/gregf-2021-12-02_21:16:13-fs-wip-cidr-blocklist-distro-basic-smithi for the fs tests

**Comment by gregsfortytwo** [comment]
Ah, this is actually busting mds client eviction as it stands. The MDS has client addresses that look like ```172.21.15.18:0/1308057869``` and so the monitor parses it as a CIDR range with an invalid bitmask.

It's probably simplest to make the UI a "mon add range" instead of overloading "mon add"; I'll try that out.

**Comment by gregsfortytwo** [comment]
Finally got shaman to build without weird system failures, so scheduled tests again:
fs http://pulpito.front.sepia.ceph.com:80/gregf-2021-12-14_19:49:56-fs-wip-cidr-blocklist-1214-2-distro-default-smithi/
rados http://pulpito.front.sepia.ceph.com:80/gregf-2021-12-14_19:51:30-rados-wip-cidr-blocklist-1214-2-distro-default-smithi

These passed! There is a failure in one of the fs tests due to ""2021-12-14T20:24:53.732478+0000 mon.a (mon.0) 1960 : cluster [WRN] Health check failed: Reduced data availability: 2 pgs peering (PG_AVAILABILITY)" in cluster log", but that is a persistent issue in master as well.

File: src/msg/msg_types.h [hunk 1]
```diff
@@ -197,14 +197,17 @@ static inline void decode(sockaddr_storage& a,
...
+ * This also happens to work to support cidr ranges, in which
+ * case the nonce contains the netmask. It's great!
  */
 struct entity_addr_t {
   typedef enum {
     TYPE_NONE = 0,
     TYPE_LEGACY = 1,  ///< legacy msgr1 protocol (ceph jewel and older)
     TYPE_MSGR2 = 2,   ///< msgr2 protocol (new in ceph kraken)
     TYPE_ANY = 3,  ///< ambiguous
+    TYPE_CIDR = 4,
```
Comment (batrick) [inline src/msg/msg_types.h:210]: Should update `operator<<` to print CIDR `entity_addr_t`  appropriately. But how should it look? A nonce already looks like a CIDR network.

File: src/msg/msg_types.h [hunk 2]
```diff
@@ -245,6 +249,8 @@ struct entity_addr_t {
   bool is_legacy() const { return type == TYPE_LEGACY; }
   bool is_msgr2() const { return type == TYPE_MSGR2; }
   bool is_any() const { return type == TYPE_ANY; }
+  // this isn't a guarantee; some client addrs will match it
+  bool maybe_cidr() const { return get_port() == 0 && nonce != 0; }
```
Comment (batrick) [inline src/msg/msg_types.h:253]: There are other solutions here that won't be ambiguous:

For IPv4: check that the first octet is `224` (multicast). If so, use the port as that octet and nonce as CIDR block range.
For IPv6: check that the first octet is `ff00` (also multicast). As with IPv4, use the port as that octet and nonce as CIDR block range.

File: doc/rados/operations/control.rst [hunk 3]
```diff
@@ -243,18 +243,23 @@ Deployments utilizing Nautilus (or later revisions of Luminous and Mimic)
...
+including the ``range`` keyword.
 
 These commands are mostly only useful for failure testing, as
 blocklists are normally maintained automatically and shouldn't need
 manual intervention. ::
 
-	ceph osd blocklist add ADDRESS[:source_port] [TIME]
-	ceph osd blocklist rm ADDRESS[:source_port]
+	ceph osd blocklist [range] add ADDRESS[:source_port][/netmask_bits] [TIME]
+	ceph osd blocklist [range] rm ADDRESS[:source_port][/netmask_bits]
```
Comment (batrick) [inline doc/rados/operations/control.rst:262]:
```suggestion
	ceph osd blocklist ["range"] add ADDRESS[:source_port][/netmask_bits] [TIME]
	ceph osd blocklist ["range"] rm ADDRESS[:source_port][/netmask_bits]
```

**Review by batrick — CHANGES_REQUESTED** [review]
my review so far; will finish tomorrow

The commit organization makes this a little more difficult. There is a `is_cidr` method which is not in the final diff. Some commits look like they could have been squashed.

File: src/osd/OSDMap.cc [hunk 4]
```diff
@@ -1358,9 +1358,72 @@ void OSDMap::set_epoch(epoch_t e)
...
+}
+
+void OSDMap::range_bits::parse(const entity_addr_t& addr) {
+  // parse it into meaningful data
+  if (addr.is_ipv6()) {
+    get_ipv6_bytes(addr.in6_addr().sin6_addr.s6_addr,
+		   &bits.ipv6.upper_64_bits, &bits.ipv6.lower_64_bits);
+    int32_t lower_shift = std::min(128-
+				   static_cast<int32_t>(addr.get_nonce()), 64);
+    int32_t upper_shift = std::max(64- //(128-b.first.get_nonce())-64
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/osd/OSDMap.cc:1386]: Oh that's just displaying a longer-form version of the calculation to illustrate where 64 came from.

File: src/msg/msg_types.h (hunk 1 above)
Comment (gregsfortytwo) [reply src/msg/msg_types.h:210]: operator<< includes the type, so it'll display as "cidr: 10.2.1.100/24" or whatever. That seems good to me?

File: src/msg/msg_types.h (hunk 2 above)
Comment (gregsfortytwo) [reply src/msg/msg_types.h:253]: Those suggestions are way more intrusive to the logic of entity_addr_t — at that point I'd probably create a new type.

In the final logic this is just used by OSDMonitor to make sure the parsed address makes sense as a CIDR address, at which point we set TYPE_CIDR and it's unambiguous.

Comment (batrick) [reply src/msg/msg_types.h:253]: Got it!

File: src/osd/OSDMap.h [hunk 5]
```diff
@@ -584,8 +584,31 @@ class OSDMap {
...
+    union {
+      ip6 ipv6;
+      ip4 ipv4;
+    } bits;
+    bool ipv6;
+    static void get_ipv6_bytes(unsigned const char *addr,
+			uint64_t *upper, uint64_t *lower);
+  public:
+    range_bits();
+    range_bits(const entity_addr_t& addr);
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/osd/OSDMap.h:605]: Unless I misunderstand that would be slower — ```emplace``` takes a key and *arguments* to construct the value, but passing ```range_bits(addr)``` would involve creating a range_bits and then invoking the copy constructor. I guess I could have used ```insert()``` and done move constructors to try and keep it efficient, but that seemed like more work.

File: src/test/osd/TestOSDMap.cc [hunk 6]
```diff
@@ -2073,6 +2073,126 @@ TEST_P(OSDMapTest, BUG_51842) {
...
+    addr.parse(a);
+    addr.set_type(entity_addr_t::TYPE_LEGACY);
+    ASSERT_TRUE(osdmap.is_blocklisted(addr, g_ceph_context));
+  }
+  for (const auto& a: unblocked_ip_addrs) {
+    entity_addr_t addr;
+    addr.parse(a);
+    addr.set_type(entity_addr_t::TYPE_LEGACY);
+    ASSERT_FALSE(osdmap.is_blocklisted(addr, g_ceph_context));
+  }
```
Comment (batrick) [inline src/test/osd/TestOSDMap.cc:2112]: This test should stop here. I'd like to see this broken up into separate tests that do:

* Test current blocklist behavior 1 address matching among many blocklisted addresses
* Test current blocklist behavior 1 address not matching
* Test current blocklist behavior with IP w/o nonce/port
* Test removing a blocklist item
* Test different synonymous CIDR ranges like: `2001::fffe/128` and `2001::0:fffe/128`. Ditto for ipv6 to match against.
* Test CIDR edge-cases (/128 and /0)

**Review by batrick — CHANGES_REQUESTED** [review]
Otherwise LGTM

File: src/osd/OSDMap.h [hunk 7]
```diff
@@ -696,8 +719,8 @@ class OSDMap {
   const utime_t& get_created() const { return created; }
   const utime_t& get_modified() const { return modified; }
 
-  bool is_blocklisted(const entity_addr_t& a) const;
-  bool is_blocklisted(const entity_addrvec_t& a) const;
+  bool is_blocklisted(const entity_addr_t& a, CephContext *cct=NULL) const;
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/osd/OSDMap.h:722]: Swapping

File: doc/rados/operations/control.rst (hunk 3 above)
Comment (gregsfortytwo) [reply doc/rados/operations/control.rst:262]: Doing

File: src/osd/OSDMap.h [hunk 8]
```diff
@@ -397,6 +397,8 @@ class OSDMap {
 
     mempool::osdmap::map<entity_addr_t,utime_t> new_blocklist;
     mempool::osdmap::vector<entity_addr_t> old_blocklist;
+    mempool::osdmap::map<entity_addr_t,utime_t> new_range_blocklist;
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/osd/OSDMap.h:400]: It's defensive.

File: src/mon/MonCommands.h [hunk 9]
```diff
@@ -1016,6 +1016,7 @@ COMMAND("osd new "
         "Reads secrets from JSON file via `-i <file>` (see man page).",
         "osd", "rw")
 COMMAND("osd blocklist "
+	"name=range,type=CephString,goodchars=[range],req=false "
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/mon/MonCommands.h:1019]: Me too, but then the format turns into
```ceph osd blocklist add ADDRESS --range```. :(

I couldn't figure out a simpler way to get it to be this optional word in the middle of the command.

File: src/osd/OSDMap.cc [hunk 10]
```diff
@@ -1336,9 +1358,72 @@ void OSDMap::set_epoch(epoch_t e)
...
+    };
+
+    bits.ipv6.lower_mask = get_mask(lower_shift);
+    bits.ipv6.upper_mask = get_mask(upper_shift);
+    ipv6 = true;
+  } else if (addr.is_ipv4()) {
+    bits.ipv4.ip_32_bits = ntohl(addr.in4_addr().sin_addr.s_addr);
+    bits.ipv4.mask = UINT32_MAX << (32-addr.get_nonce());
+  } else {
+    // uh...
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/osd/OSDMap.cc:1403]: I've gotten allergic to that lately after the number of defensive asserts I had that ended up triggering on valid state due to insufficient imagination on my part... :/

File: src/test/osd/TestOSDMap.cc [hunk 11]
```diff
@@ -2073,6 +2073,126 @@ TEST_P(OSDMapTest, BUG_51842) {
     }
 }
 
+TEST_F(OSDMapTest, blocklisting) {
+  set_up_map(6); //whatever
+
+  // I pulled the first two ranges and their start/end points from
+  // https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing#CIDR_notation
+  string range_addrs[] = {"198.51.100.0/22", "2001:db8::/48", "3001:db8::/72"};
```
[the remark this answers is not in the document]
Comment (gregsfortytwo) [reply src/test/osd/TestOSDMap.cc]: These are added.

File: src/test/osd/TestOSDMap.cc (hunk 6 above)
Comment (gregsfortytwo) [reply src/test/osd/TestOSDMap.cc:2131]: I split the tests up some (thought not all the way) and I think we have all the edge cases you've mentioned here now.

**Comment by gregsfortytwo** [comment]
Jenkins test api

**Comment by ljflores** [comment]
@gregsfortytwo looks like there is a regression related to this PR in the rados suite. I created a Tracker issue for it; can you take a look? https://tracker.ceph.com/issues/55419

http://pulpito.front.sepia.ceph.com/yuriw-2022-04-22_13:56:48-rados-wip-yuri2-testing-2022-04-22-0500-distro-default-smithi/6800292/
```
2022-04-22T14:23:30.757 INFO:tasks.workunit.client.0.smithi085.stderr:listed 5 entries
2022-04-22T14:23:30.772 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  return 0
2022-04-22T14:23:30.773 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1429: test_mon_osd:  expect_false 'ceph osd blocklist add 192.168.0.1/-1'
2022-04-22T14:23:30.773 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:35: expect_false:  set -x
2022-04-22T14:23:30.773 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  'ceph osd blocklist add 192.168.0.1/-1'
2022-04-22T14:23:30.773 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh: line 36: ceph osd blocklist add 192.168.0.1/-1: No such file or directory
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  return 0
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1430: test_mon_osd:  expect_false 'ceph osd blocklist add 192.168.0.1/foo'
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:35: expect_false:  set -x
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  'ceph osd blocklist add 192.168.0.1/foo'
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh: line 36: ceph osd blocklist add 192.168.0.1/foo: No such file or directory
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  return 0
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1433: test_mon_osd:  expect_false 'ceph osd blocklist add 1234.56.78.90/100'
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:35: expect_false:  set -x
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  'ceph osd blocklist add 1234.56.78.90/100'
2022-04-22T14:23:30.774 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh: line 36: ceph osd blocklist add 1234.56.78.90/100: No such file or directory
2022-04-22T14:23:30.775 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:36: expect_false:  return 0
2022-04-22T14:23:30.775 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1436: test_mon_osd:  bl=192.168.0.1/24
2022-04-22T14:23:30.775 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1437: test_mon_osd:  ceph osd blocklist range add 192.168.0.1/24
2022-04-22T14:23:32.991 INFO:tasks.workunit.client.0.smithi085.stderr:blocklisting cidr:192.168.0.1:0/24 until 2022-04-22T15:23:32.109570+0000 (3600 sec)
2022-04-22T14:23:33.004 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1438: test_mon_osd:  ceph osd blocklist ls
2022-04-22T14:23:33.004 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1438: test_mon_osd:  grep 192.168.0.1/24
2022-04-22T14:23:33.427 INFO:tasks.workunit.client.0.smithi085.stderr:listed 6 entries
2022-04-22T14:23:33.436 INFO:tasks.workunit.client.0.smithi085.stderr:/home/ubuntu/cephtest/clone.client.0/qa/workunits/cephtool/test.sh:1: test_mon_osd:  rm -fr /tmp/cephtool.dfI
2022-04-22T14:23:33.437 DEBUG:teuthology.orchestra.run:got remote process result: 1
2022-04-22T14:23:33.438 INFO:tasks.workunit:Stopping ['cephtool'] on client.0...
2022-04-22T14:23:33.438 DEBUG:teuthology.orchestra.run.smithi085:> sudo rm -rf -- /home/ubuntu/cephtest/workunits.list.client.0 /home/ubuntu/cephtest/clone.client.0
2022-04-22T14:23:33.680 ERROR:teuthology.run_tasks:Saw exception from tasks.
```

**Comment by ljflores** [comment]
Also @gregsfortytwo was there a final teuthology run done for this PR?

**Comment by gregsfortytwo** [comment]
...my apologies, I'm looking at the test history and realizing I ran it through less of the suite than I thought I had. Since the only recent changes were in the unit test code I didn't spin up another test, but I definitely should have given what it's actually run through. :/

**Comment by neha-ojha** [comment]
@gregsfortytwo I am marking https://tracker.ceph.com/issues/55419 as resolved for now. Is there a tracker this to track the backport of this feature? we should not forget to include the fix.

**Comment by gregsfortytwo** [comment]
> @gregsfortytwo I am marking https://tracker.ceph.com/issues/55419 as resolved for now. Is there a tracker this to track the backport of this feature? we should not forget to include the fix.

https://tracker.ceph.com/issues/53050 is Pending Backport; I was waiting on the test fix to make sure there weren't any other issues before I generate backport PRs. Will get to them in the next day or two.

**Pull request merged**